File: js/ISA/ISA/Fable.js
import { toString } from "../../fable_modules/fable-library.4.1.4/Types.js";
import { append } from "../../fable_modules/fable-library.4.1.4/List.js";
import { append as append_1 } from "../../fable_modules/fable-library.4.1.4/Array.js";
import { Array_distinct, List_distinct } from "../../fable_modules/fable-library.4.1.4/Seq2.js";
import { structuralHash, equals } from "../../fable_modules/fable-library.4.1.4/Util.js";

export function isMap_generic(l1) {
    let copyOfStruct;
    return ((copyOfStruct = l1, toString(copyOfStruct))).indexOf("map [") === 0;
}

export function isList_generic(l1) {
    let s;
    let copyOfStruct = l1;
    s = toString(copyOfStruct);
    if (s.indexOf("[") === 0) {
        return !(s.indexOf("seq [") === 0);
    }
    else {
        return false;
    }
}

export function append_generic(l1, l2) {
    if (l2 == null) {
        return l1;
    }
    else if (isList_generic(l1)) {
        return append(l1, l2);
    }
    else {
        return append_1(l1, l2);
    }
}

export function distinct_generic(l1) {
    if (isList_generic(l1)) {
        return List_distinct(l1, {
            Equals: equals,
            GetHashCode: structuralHash,
        });
    }
    else {
        return Array_distinct(l1, {
            Equals: equals,
            GetHashCode: structuralHash,
        });
    }
}

File: src/ISA/ISA/Fable.fs
[<RequireQualifiedAccess>]
module ARCtrl.ISA.Fable

open Fable.Core

#if FABLE_COMPILER_PYTHON
open Fable.Core.PyInterop
#else
open Fable.Core.JsInterop
#endif

[<Emit("console.log($0)")>]
let print (msg:obj) : unit = 
    printfn "%O" msg

let isMap_generic l1 = l1.ToString().StartsWith("map [")

let isList_generic l1 = 
    let s = l1.ToString()
    printfn "%A" (s)
    s.StartsWith("[") && (s.StartsWith "seq [" |> not)

let append_generic l1 l2 =
    // This isNull check is necessary because in the API.update functionality we only check the type of l1. 
    // There l1 can be a sequence (Some sequence in dotnet) and l2 would be an undefined (None in dotnet).
    // We need to check if l2 is null and if so return l1.
    if !!isNull l2 then l1

    else
        if isList_generic l1 then 
            !!List.append l1 l2
        else
            !!Array.append l1 l2

let distinct_generic l1 =       
    if isList_generic l1 then 
        !!List.distinct l1
    else
        !!Array.distinct l1

let hashSeq (s:seq<'a>) = 
    s
    |> Seq.map (fun x -> x.GetHashCode())
    |> Seq.reduce (fun a b -> a + b)